## Importing important Libraries and Classifiers

In [4]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.ensemble import RandomForestClassifier
from sklearn import svm
from sklearn.naive_bayes import GaussianNB
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import train_test_split
from sklearn import metrics
from sklearn.preprocessing import LabelEncoder,StandardScaler

## Loading train.csv and test.csv data

In [180]:
train_df = pd.read_csv('train.csv')
test_df = pd.read_csv('test.csv')

## Exploring Data

In [181]:
pd.set_option('display.max_columns',200)
pd.set_option('display.max_rows',1000)

In [182]:
train_df.head()

id                species   margin1   margin2   margin3   margin4  \
0   1            Acer_Opalus  0.007812  0.023438  0.023438  0.003906   
1   2  Pterocarya_Stenoptera  0.005859  0.000000  0.031250  0.015625   
2   3   Quercus_Hartwissiana  0.005859  0.009766  0.019531  0.007812   
3   5        Tilia_Tomentosa  0.000000  0.003906  0.023438  0.005859   
4   6     Quercus_Variabilis  0.005859  0.003906  0.048828  0.009766   

    margin5   margin6   margin7  margin8   margin9  margin10  margin11  \
0  0.011719  0.009766  0.027344      0.0  0.001953  0.033203  0.013672   
1  0.025391  0.001953  0.019531      0.0  0.000000  0.007812  0.003906   
2  0.003906  0.005859  0.068359      0.0  0.000000  0.044922  0.007812   
3  0.021484  0.019531  0.023438      0.0  0.013672  0.017578  0.001953   
4  0.013672  0.015625  0.005859      0.0  0.000000  0.005859  0.001953   

   margin12  margin13  margin14  margin15  margin16  margin17  margin18  \
0  0.019531  0.066406  0.000000  0.029297       0.0  0.031250  0.011719   
1  0.027344  0.023438  0.000000  0.033203       0.0  0.009766  0.009766   
2  0.011719  0.021484  0.001953  0.025391       0.0  0.009766  0.011719   
3  0.019531  0.001953  0.003906  0.035156       0.0  0.005859  0.000000   
4  0.044922  0.041016  0.011719  0.041016       0.0  0.009766  0.015625   

   margin19  margin20  margin21  margin22  margin23  margin24  margin25  \
0  0.000000  0.025391  0.023438  0.001953       0.0  0.015625  0.000000   
1  0.007812  0.007812  0.019531  0.007812       0.0  0.000000  0.007812   
2  0.007812  0.005859  0.041016  0.001953       0.0  0.015625  0.000000   
3  0.001953  0.003906  0.039062  0.009766       0.0  0.009766  0.005859   
4  0.011719  0.007812  0.021484  0.000000       0.0  0.000000  0.015625   

   margin26  margin27  margin28  margin29  margin30  margin31  margin32  \
0  0.031250  0.000000  0.013672  0.029297  0.015625  0.011719  0.003906   
1  0.027344  0.003906  0.037109  0.007812  0.048828  0.054688  0.027344   
2  0.009766  0.001953  0.009766  0.009766  0.015625  0.005859  0.000000   
3  0.027344  0.001953  0.017578  0.003906  0.021484  0.027344  0.027344   
4  0.009766  0.001953  0.017578  0.041016  0.005859  0.021484  0.013672   

   margin33  margin34  margin35  margin36  margin37  margin38  margin39  \
0  0.025391  0.000000  0.001953  0.011719  0.009766  0.041016  0.037109   
1  0.003906  0.000000  0.000000  0.003906  0.013672  0.033203  0.033203   
2  0.017578  0.007812  0.005859  0.009766  0.019531  0.042969  0.021484   
3  0.000000  0.003906  0.011719  0.003906  0.005859  0.044922  0.056641   
4  0.017578  0.000000  0.005859  0.009766  0.027344  0.083984  0.027344   

   margin40  margin41  margin42  margin43  margin44  margin45  margin46  \
0  0.019531  0.000000  0.009766  0.021484  0.015625  0.007812  0.013672   
1  0.019531  0.031250  0.009766  0.007812  0.031250  0.001953  0.039062   
2  0.001953  0.000000  0.005859  0.015625  0.009766  0.011719  0.011719   
3  0.027344  0.015625  0.007812  0.017578  0.029297  0.001953  0.042969   
4  0.027344  0.001953  0.001953  0.017578  0.007812  0.009766  0.011719   

   margin47  margin48  margin49  margin50  margin51  margin52  margin53  \
0  0.027344  0.062500  0.000000  0.017578  0.031250       0.0  0.044922   
1  0.029297  0.031250  0.035156  0.000000  0.007812       0.0  0.046875   
2  0.031250  0.109380  0.000000  0.046875  0.031250       0.0  0.035156   
3  0.085938  0.037109  0.029297  0.007812  0.013672       0.0  0.052734   
4  0.080078  0.039062  0.009766  0.000000  0.029297       0.0  0.023438   

   margin54  margin55  margin56  margin57  margin58  margin59  margin60  \
0  0.007812  0.025391  0.003906  0.013672  0.015625  0.013672  0.003906   
1  0.046875  0.029297  0.009766  0.017578  0.007812  0.013672  0.019531   
2  0.000000  0.105470  0.000000  0.015625  0.015625  0.019531  0.001953   
3  0.031250  0.037109  0.017578  0.003906  0.000000  0.001953  0.003906   
4  0.021484  0.003906  0.017578  0.015625  0

In [183]:
test_df.head()

id   margin1   margin2   margin3   margin4   margin5   margin6   margin7  \
0   4  0.019531  0.009766  0.078125  0.011719  0.003906  0.015625  0.005859   
1   7  0.007812  0.005859  0.064453  0.009766  0.003906  0.013672  0.007812   
2   9  0.000000  0.000000  0.001953  0.021484  0.041016  0.000000  0.023438   
3  12  0.000000  0.000000  0.009766  0.011719  0.017578  0.000000  0.003906   
4  13  0.001953  0.000000  0.015625  0.009766  0.039062  0.000000  0.009766   

   margin8   margin9  margin10  margin11  margin12  margin13  margin14  \
0      0.0  0.005859  0.023438  0.005859  0.021484  0.076172  0.001953   
1      0.0  0.033203  0.023438  0.009766  0.019531  0.039062  0.027344   
2      0.0  0.011719  0.005859  0.001953  0.021484  0.001953  0.019531   
3      0.0  0.003906  0.001953  0.000000  0.029297  0.000000  0.039062   
4      0.0  0.005859  0.000000  0.001953  0.033203  0.000000  0.003906   

   margin15  margin16  margin17  margin18  margin19  margin20  margin21  \
0  0.035156  0.000000  0.001953  0.021484  0.001953  0.015625  0.009766   
1  0.017578  0.000000  0.017578  0.015625  0.009766  0.009766  0.042969   
2  0.035156  0.000000  0.007812  0.001953  0.046875  0.003906  0.003906   
3  0.037109  0.003906  0.007812  0.005859  0.007812  0.005859  0.025391   
4  0.011719  0.000000  0.003906  0.005859  0.011719  0.001953  0.035156   

   margin22  margin23  margin24  margin25  margin26  margin27  margin28  \
0  0.000000  0.000000  0.005859  0.007812  0.015625  0.000000  0.017578   
1  0.007812  0.000000  0.003906  0.005859  0.011719  0.007812  0.009766   
2  0.013672  0.015625  0.037109  0.027344  0.009766  0.023438  0.027344   
3  0.023438  0.015625  0.021484  0.033203  0.009766  0.050781  0.015625   
4  0.027344  0.000000  0.003906  0.056641  0.013672  0.029297  0.023438   

   margin29  margin30  margin31  margin32  margin33  margin34  margin35  \
0  0.029297  0.021484  0.001953  0.003906  0.025391  0.000000  0.013672   
1  0.037109  0.013672  0.005859  0.003906  0.019531  0.000000  0.005859   
2  0.005859  0.011719  0.046875  0.056641  0.001953  0.003906  0.003906   
3  0.005859  0.005859  0.009766  0.015625  0.001953  0.003906  0.001953   
4  0.000000  0.013672  0.074219  0.083984  0.009766  0.000000  0.000000   

   margin36  margin37  margin38  margin39  margin40  margin41  margin42  \
0  0.005859  0.042969  0.064453  0.031250  0.005859  0.003906  0.001953   
1  0.029297  0.019531  0.041016  0.005859  0.003906  0.001953  0.009766   
2  0.005859  0.001953  0.003906  0.013672  0.007812  0.058594  0.021484   
3  0.005859  0.001953  0.039062  0.027344  0.009766  0.001953  0.000000   
4  0.001953  0.007812  0.031250  0.009766  0.023438  0.033203  0.005859   

   margin43  margin44  margin45  margin46  margin47  margin48  margin49  \
0  0.021484  0.025391  0.005859  0.001953  0.023438  0.068359  0.000000   
1  0.031250  0.001953  0.029297  0.007812  0.037109  0.009766  0.001953   
2  0.041016  0.015625  0.007812  0.025391  0.013672  0.017578  0.042969   
3  0.031250  0.015625  0.001953  0.029297  0.027344  0.068359  0.021484   
4  0.011719  0.019531  0.000000  0.029297  0.048828  0.009766  0.072266   

   margin50  margin51  margin52  margin53  margin54  margin55  margin56  \
0  0.011719  0.031250  0.001953  0.011719  0.003906  0.007812  0.000000   
1  0.058594  0.029297  0.001953  0.021484  0.003906  0.005859  0.000000   
2  0.003906  0.003906  0.007812  0.042969  0.017578  0.017578  0.001953   
3  0.046875  0.007812  0.039062  0.025391  0.011719  0.025391  0.007812   
4  0.013672  0.000000  0.000000  0.025391  0.046875  0.007812  0.027344   

   margin57  margin58  margin59  margin60  margin61  margin62  margin63  \
0  0.009766  0.003906  0.054688  0.000000  0.000000  0.000000  0.052734   
1  0.021484  0.015625  0.048828  0.000000  0.000000  0.007812  0.029297   
2  0.000000  0.005859  0.001953  0.039062  0.003906  0.029297  0.000000   
3  0.005859  0.000000  0.011719  0.023438  0.003906  0.044922  0.007812   
4 

In [184]:
train_df.shape

(990, 194)

In [185]:
train_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 990 entries, 0 to 989
Columns: 194 entries, id to texture64
dtypes: float64(192), int64(1), object(1)
memory usage: 1.5+ MB


In [186]:
train_df.dtypes

id             int64
species       object
margin1      float64
margin2      float64
margin3      float64
margin4      float64
margin5      float64
margin6      float64
margin7      float64
margin8      float64
margin9      float64
margin10     float64
margin11     float64
margin12     float64
margin13     float64
margin14     float64
margin15     float64
margin16     float64
margin17     float64
margin18     float64
margin19     float64
margin20     float64
margin21     float64
margin22     float64
margin23     float64
margin24     float64
margin25     float64
margin26     float64
margin27     float64
margin28     float64
margin29     float64
margin30     float64
margin31     float64
margin32     float64
margin33     float64
margin34     float64
margin35     float64
margin36     float64
margin37     float64
margin38     float64
margin39     float64
margin40     float64
margin41     float64
margin42     float64
margin43     float64
margin44     float64
margin45     float64
margin46     

In [187]:
test_df.shape

(594, 193)

In [188]:
test_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 594 entries, 0 to 593
Columns: 193 entries, id to texture64
dtypes: float64(192), int64(1)
memory usage: 895.7 KB


In [189]:
## Droping ID as it is not useful for classification
train_df = train_df.drop('id',axis=1)
test_df = test_df.drop('id',axis=1)

In [190]:
train_df.head(2)

species   margin1   margin2   margin3   margin4   margin5  \
0            Acer_Opalus  0.007812  0.023438  0.023438  0.003906  0.011719   
1  Pterocarya_Stenoptera  0.005859  0.000000  0.031250  0.015625  0.025391   

    margin6   margin7  margin8   margin9  margin10  margin11  margin12  \
0  0.009766  0.027344      0.0  0.001953  0.033203  0.013672  0.019531   
1  0.001953  0.019531      0.0  0.000000  0.007812  0.003906  0.027344   

   margin13  margin14  margin15  margin16  margin17  margin18  margin19  \
0  0.066406       0.0  0.029297       0.0  0.031250  0.011719  0.000000   
1  0.023438       0.0  0.033203       0.0  0.009766  0.009766  0.007812   

   margin20  margin21  margin22  margin23  margin24  margin25  margin26  \
0  0.025391  0.023438  0.001953       0.0  0.015625  0.000000  0.031250   
1  0.007812  0.019531  0.007812       0.0  0.000000  0.007812  0.027344   

   margin27  margin28  margin29  margin30  margin31  margin32  margin33  \
0  0.000000  0.013672  0.029297  0.015625  0.011719  0.003906  0.025391   
1  0.003906  0.037109  0.007812  0.048828  0.054688  0.027344  0.003906   

   margin34  margin35  margin36  margin37  margin38  margin39  margin40  \
0       0.0  0.001953  0.011719  0.009766  0.041016  0.037109  0.019531   
1       0.0  0.000000  0.003906  0.013672  0.033203  0.033203  0.019531   

   margin41  margin42  margin43  margin44  margin45  margin46  margin47  \
0   0.00000  0.009766  0.021484  0.015625  0.007812  0.013672  0.027344   
1   0.03125  0.009766  0.007812  0.031250  0.001953  0.039062  0.029297   

   margin48  margin49  margin50  margin51  margin52  margin53  margin54  \
0   0.06250  0.000000  0.017578  0.031250       0.0  0.044922  0.007812   
1   0.03125  0.035156  0.000000  0.007812       0.0  0.046875  0.046875   

   margin55  margin56  margin57  margin58  margin59  margin60  margin61  \
0  0.025391  0.003906  0.013672  0.015625  0.013672  0.003906  0.005859   
1  0.029297  0.009766  0.017578  0.007812  0.013672  0.019531  0.000000   

   margin62  margin63  margin64    shape1    shape2    shape3    shape4  \
0  0.003906  0.019531  0.001953  0.000647  0.000609  0.000576  0.000553   
1  0.000000  0.003906  0.000000  0.000749  0.000695  0.000720  0.000709   

     shape5    shape6    shape7    shape8    shape9   shape10   shape11  \
0  0.000516  0.000496  0.000474  0.000453  0.000454  0.000429  0.000418   
1  0.000688  0.000660  0.000624  0.000585  0.000556  0.000531  0.000506   

    shape12   shape13   shape14   shape15   shape16   shape17   shape18  \
0  0.000410  0.000406  0.000398  0.000399  0.000389  0.000399  0.000416   
1  0.000484  0.000467  0.000453  0.000442  0.000435  0.000435  0.000433   

    shape19  shape20   shape21   shape22   shape23   shape24   shape25  \
0  0.000414  0.00043  0.000422  0.000437  0.000452  0.000467  0.000470   
1  0.000443  0.00046  0.000471  0.000486  0.000512  0.000538  0.000567   

    shape26   shape27   shape28   shape29   shape30   shape31   shape32  \
0  0.000484  0.000489  0.000490  0.000475  0.000486  0.000484  0.000500   
1  0.000605  0.000644  0.000687  0.000731  0.000777  0.000832  0.000884   

    shape33   shape34   shape35   shape36   shape37   shape38   shape39  \
0  0.000513  0.000511  0.000519  0.000513  0.000502  0.000498  0.000487   
1  0.000902  0.000862  0.000842  0.000783  0.000731  0.000679  0.000639   

    shape40   shape41   shape42   shape43   shape44   shape45   shape46  \
0  0.000471  0.000458  0.000440  0.000436  0.000427  0.000394  0.000385   
1  0.000596  0.000560  0.000528  0.000496  0.000474  0.000462  0.000447   

    shape47   shape48   shape49   shape50   shape51   shape52   shape53  \
0  0.000382  0.000370  0.000367  0.000373  0.000386  0.000389  0.000391   
1  0.000434  0.000433  0.000435  0.000436  0.000447  0.000464  0.000479   

    shape54   shape55   shape56   shape57   shape58   shape59   shape60  \
0  0.000414  0.000422  0.000434  0.000452  0.000471  0.000485  0.000512   
1  0.000496  0.000518 

In [191]:
## Seperating Feature and Target data
X = train_df.iloc[:,1:]
Y = train_df['species']

In [192]:
X.shape

(990, 192)

In [193]:
Y.shape

(990,)

## Encoding Target string data using LabelEncoder

In [194]:
def encoder(data):
    le = LabelEncoder().fit(data)
    encoded_data = le.transform(data)
    classes = le.classes_
    return encoded_data,classes,le   

In [195]:
Y,classes,encod_model = encoder(Y)

In [196]:
encoded_number = np.arange(len(classes))
classes_df = pd.DataFrame({'Species':classes})
classes_df.head()

Species
0  Acer_Capillipes
1  Acer_Circinatum
2        Acer_Mono
3      Acer_Opalus
4    Acer_Palmatum

## Extract the values from train set by stratifying them and dividing it into 80:20 ratio

In [197]:
X_train,X_test,Y_train,Y_test = train_test_split(X,Y,test_size = 0.2, random_state = 85)

In [198]:
X_train.shape

(792, 192)

In [199]:
Y_train.shape

(792,)

In [200]:
X_test.shape

(198, 192)

In [201]:
Y_test.shape

(198,)

## Applying all 4 (Randomforest, SVM, NaiveBayes, DecisionTrees) classifiers to find the best one

In [202]:
# Creating the classifiers class
class Classifier:
    def __init__(self,model):
        self.model = model
        self.pred = None
        self.accuracy = None
        self.loss = None
    
    def fit(self,x_train,y_train):
        self.model.fit(x_train,y_train)
    def predict(self,test):
        self.pred = self.model.predict(test)
        return self.pred
    
    def accuracy_score(self,actual):
        self.accuracy = (metrics.accuracy_score(actual,self.pred))*100
        print(self.accuracy)

## Applying DecisionTree

In [203]:
DT = Classifier(DecisionTreeClassifier(random_state=50))

In [204]:
DT.fit(X_train,Y_train)

In [205]:
DT.predict(X_test)

array([76, 85, 96, 29, 37,  5, 57, 77, 84,  2, 88, 66, 67, 53, 74, 55, 38,
       95, 32, 65, 76, 46, 71, 39, 29, 80, 20, 45, 52, 21, 40, 51, 20, 71,
       28,  4, 11, 76, 39,  8, 88, 54, 20, 81, 32, 59, 60, 30, 32, 96, 13,
       62,  5, 23, 84, 84, 79, 60, 72,  7, 24, 47, 78, 74, 83, 63, 47,  6,
       42, 16, 37, 66, 75, 11, 60, 38, 90, 61, 38, 45, 27, 59, 87, 20, 79,
        6, 96, 11, 44, 80, 60,  9, 84, 75, 43, 40, 93, 43,  9, 56, 52, 57,
       20, 76,  5, 94, 38, 43, 97, 63, 27, 72, 61,  0, 14,  3, 71, 87, 27,
        6, 80, 55, 31, 37, 93, 36, 50, 12, 69, 88, 55, 13, 88, 85, 56, 79,
       17, 57, 43, 21,  5,  2, 78, 49, 88, 30, 54,  8, 70, 68, 24, 65, 26,
       93,  5, 45, 92, 30, 58, 97, 93, 40, 82, 44, 64, 21, 54, 63, 67, 14,
       27, 62, 45, 26,  8, 28, 47, 53, 42, 53, 25, 45, 36, 95, 55, 41, 70,
       98, 80, 64, 76, 75, 24, 57, 64, 21, 81,  0], dtype=int64)

In [206]:
DT.accuracy_score(Y_test)

69.6969696969697


### Using Decision Tree Accuracy is 69.7 %

## Applying RandomForest classifier

In [207]:
RFC = Classifier(RandomForestClassifier(n_estimators=50,random_state=59))

In [208]:
RFC.fit(X_train,Y_train)

In [209]:
RFC.predict(X_test)

array([85, 85, 23, 29, 43,  5, 57, 77,  7,  2, 66, 56, 67, 53, 51, 55, 37,
       97, 84, 50, 76, 46, 63, 39, 29, 71, 53, 60, 52, 21, 40, 51, 20, 80,
       28,  4, 11, 76, 39,  8, 21, 46, 83, 86, 32, 59, 60, 30, 32,  3, 13,
       62,  5, 23, 92, 84, 70, 53, 72,  7, 71, 47, 78, 74, 83, 63, 47,  6,
       42, 16, 37, 66, 75, 11, 60, 38, 90, 61, 37, 45, 27, 59, 87, 55, 67,
        6, 96, 11, 44, 63, 52,  9, 92, 74, 67, 40, 93, 43,  9, 56, 52, 57,
       70, 76, 66, 17, 87, 90, 97, 71, 27, 72, 61,  0, 14, 86, 71, 87, 27,
        6, 80, 53, 31, 37, 93, 36, 50, 12, 69, 83, 64, 13, 88, 85, 78,  3,
       17, 57, 28, 21,  5,  2, 72, 49, 88, 30, 64,  8, 90, 68, 24, 69, 26,
       93,  5, 45, 53, 30, 58, 97, 93, 40, 82, 44, 64, 21, 54, 80, 67, 14,
       75, 79, 45, 48, 25, 28, 47, 96, 27, 53, 25, 45, 36, 11, 55, 41, 70,
       98, 80, 56, 76, 75, 24, 57, 64, 21, 81,  0], dtype=int64)

In [210]:
RFC.accuracy_score(Y_test)

96.96969696969697


### Using RandomForestClassifier Accuracy is 96.97 %

## Applying SVM classifier

In [211]:
vector = Classifier(svm.LinearSVC())

In [212]:
vector.fit(X_train,Y_train)

In [213]:
vector.predict(X_test)

array([69, 85, 23, 29, 43,  5, 57, 77,  7,  2, 66, 56, 67, 50, 51, 55, 37,
       97, 84, 50, 76, 46, 63, 39, 29, 71, 48, 60, 52, 21, 40, 51, 20, 71,
       30,  4, 95, 76, 39,  8, 73, 46, 83, 86, 32, 59, 60, 30, 32, 83, 13,
       62,  5, 23, 92, 84, 72, 53, 72,  7, 71, 47, 78, 32, 52, 63, 47, 94,
       42, 16, 37, 66, 75, 11, 60, 38, 90, 61, 37, 94, 27, 59, 87, 55, 67,
        6, 96, 95, 44, 63, 52,  9, 92, 51, 67, 33, 93, 43,  9, 56, 52, 57,
       59, 76, 66, 17, 87, 90, 97, 63, 60, 72, 61,  0, 14, 86, 71, 87, 66,
        6, 71, 53, 68, 37, 93, 62, 50, 12, 50, 83, 64, 13, 88, 85, 78,  3,
       17, 57, 15, 68,  5,  2, 72, 49, 88, 42, 96,  8, 90, 68, 24, 52, 59,
       20,  5, 45, 52, 30, 58, 97, 93, 33, 82, 44, 64, 21, 54, 71, 22, 14,
       75, 79, 45, 48, 25, 28, 47, 96, 27, 50, 25, 45, 29, 16, 55, 41, 70,
       98, 71, 56, 76, 75, 24, 57, 32, 34, 81,  0], dtype=int64)

In [214]:
vector.accuracy_score(Y_test)

79.29292929292929


### Using LinearSVC classifier Accuracy is 79.29 %

## Applying NaiveBayes Classifier

In [215]:
NB = Classifier(GaussianNB())

In [216]:
NB.fit(X_train,Y_train)

In [217]:
NB.predict(X_test)

array([69, 85, 23, 29, 96, 78, 57, 77,  7,  2, 66, 56, 36, 12, 56, 55, 39,
       77, 84, 50, 76, 46, 71, 70, 79, 71, 26, 60, 52, 46, 70, 26,  0, 80,
       28,  4, 19, 85, 54,  8, 73, 56, 83, 86, 32, 59, 60, 69, 32,  3, 83,
       77, 26, 65, 92, 69, 36, 12, 19, 13, 71, 47, 78, 64, 73, 63, 19,  6,
       42, 78, 38, 66, 75, 14, 97, 38, 90, 54, 38, 45, 73, 59, 87, 54, 19,
        6, 73, 19, 44, 71, 69,  9, 92, 56, 67, 70, 93, 19,  9, 78, 52, 44,
       78, 85, 77, 17, 69, 90, 77, 78, 50, 70, 61, 20, 19, 69, 71, 87, 83,
        6, 80, 54, 69, 38, 93, 73, 50, 12, 50, 83, 56, 13, 88, 85, 78,  3,
       19, 77, 69, 46,  5,  2, 72, 49, 26, 36, 56, 16, 90,  3, 85, 69, 26,
       93, 78, 45, 53, 69, 58, 77, 93, 69, 19, 77, 56, 21, 85, 80, 67, 26,
       56, 58, 19, 14, 25, 28, 47, 43, 66, 53, 70, 84, 36, 19, 55, 26, 19,
       98, 78, 85, 76, 41, 85, 44, 56, 21, 81,  0], dtype=int64)

In [218]:
NB.accuracy_score(Y_test)

47.474747474747474


### Using NaiveBayes classifier Accuracy is 47.47 %

## RandomForestClassifier has the best Accuracy so will use this to predict import test.csv data

In [219]:
prediction = RFC.predict(test_df)

In [220]:
prediction

array([51, 50,  1, 19, 14,  3,  3, 29, 84,  8, 43, 74, 75, 10, 52, 46, 45,
       73, 13, 71, 61, 68, 57, 77,  1, 70, 28, 15, 35, 70, 53, 74, 47, 50,
        4, 36, 14, 55, 36, 93,  8, 32,  8,  9, 71, 70, 38, 23, 94, 18, 17,
        5, 55, 94, 14, 86, 62, 33, 51, 98, 88, 56, 21, 59, 65, 11, 48,  5,
       13,  4, 54, 57, 29,  7, 31, 98, 92, 84, 25, 10, 61, 43, 85, 24,  1,
        2, 23, 83, 40, 22, 48, 90, 25, 21, 37, 74, 41, 95,  7, 89, 98, 77,
        3, 12, 31, 84, 53, 96, 64, 72, 93, 93, 67, 30,  8, 88, 60, 87,  6,
       57, 34, 34, 60, 17, 75, 27, 51, 31, 39, 23, 24,  2, 41, 61, 24, 97,
       29, 28, 68, 81, 42, 51, 79, 62, 60, 52, 95, 64, 42, 96, 95, 20, 59,
       35, 86,  1, 26, 38, 43, 51, 20, 60, 46, 79, 22, 79, 55, 87, 65, 97,
       75, 21, 29, 21, 11, 10, 58, 94, 27, 22, 15, 45, 89, 54, 43,  5, 23,
       94, 40, 49, 89, 72, 36, 11, 81, 95, 18, 91, 29, 64, 80,  6, 78, 45,
       28,  9, 78, 90, 44, 89, 92, 13,  2, 59,  0, 96, 70, 32, 29, 78, 91,
       55, 44, 38,  5, 60

## Decoding the encoded output and assigning it to Test_DF

In [222]:
test_df['predicted_species'] = encod_model.inverse_transform(prediction)

In [226]:
pd.set_option('display.max_columns',10)
test_df.head()

margin1   margin2   margin3   margin4   margin5        ...          \
0  0.019531  0.009766  0.078125  0.011719  0.003906        ...           
1  0.007812  0.005859  0.064453  0.009766  0.003906        ...           
2  0.000000  0.000000  0.001953  0.021484  0.041016        ...           
3  0.000000  0.000000  0.009766  0.011719  0.017578        ...           
4  0.001953  0.000000  0.015625  0.009766  0.039062        ...           

   texture61  texture62  texture63  texture64  predicted_species  
0        0.0   0.000000   0.003906   0.053711  Quercus_Agrifolia  
1        0.0   0.000977   0.037109   0.044922     Quercus_Afares  
2        0.0   0.015625   0.000000   0.000000    Acer_Circinatum  
3        0.0   0.089844   0.000000   0.008789    Castanea_Sativa  
4        0.0   0.007812   0.009766   0.007812      Alnus_Viridis  

[5 rows x 193 columns]